server/src/main/java/io/druid/server/coordinator/rules/LoadRule.java

```diff
@@ -49,188 +51,314 @@
-    int currTotalReplicantsInCluster = totalReplicantsInCluster;
-    while (currReplicantsInTier < expectedReplicantsInTier) {
-      boolean replicate = currTotalReplicantsInCluster > 0;
+    final int numToAssign = targetReplicantsInTier - currentReplicantsInTier;
+    // if nothing to assign or no holders available for assignment
+    if (numToAssign <= 0 || holders.isEmpty()) {
+      return 0;
+    }
 
-      if (replicate && !replicationManager.canCreateReplicant(tier)) {
-        break;
+    final ReplicationThrottler throttler = params.getReplicationManager();
+    for (int numAssigned = 0; numAssigned < numToAssign; numAssigned++) {
+      if (!throttler.canCreateReplicant(tier)) {
+        return numAssigned;
       }
 
-      final ServerHolder holder = strategy.findNewSegmentHomeReplicator(segment, serverHolderList);
-
+      // Retrieves from cache if available
+      ServerHolder holder = strategyCache.remove(tier);
+      // Does strategy call if not in cache
       if (holder == null) {
-        log.warn(
-            "Not enough [%s] servers or node capacity to assign segment[%s]! Expected Replicants[%d]",
-            tier,
-            segment.getIdentifier(),
-            expectedReplicantsInTier
-        );
-        break;
+        holder = params.getBalancerStrategy().findNewSegmentHomeReplicator(segment, holders);
       }
 
-      if (replicate) {
-        replicationManager.registerReplicantCreation(
-            tier, segment.getIdentifier(), holder.getServer().getHost()
+      if (holder == null) {
+        log.warn(
+            "No available [%s] servers or node capacity to assign segment[%s]! Expected Replicants[%d]",
+            tier, segment.getIdentifier(), targetReplicantsInTier
         );
+        return numAssigned;
       }
+      holders.remove(holder);
 
-      holder.getPeon().loadSegment(
-          segment,
-          new LoadPeonCallback()
-          {
-            @Override
-            public void execute()
-            {
-              replicationManager.unregisterReplicantCreation(
-                  tier,
-                  segment.getIdentifier(),
-                  holder.getServer().getHost()
-              );
-            }
-          }
-      );
-
-      ++assignedCount;
-      ++currReplicantsInTier;
-      ++currTotalReplicantsInCluster;
+      final String segmentId = segment.getIdentifier();
+      final String holderHost = holder.getServer().getHost();
+      throttler.registerReplicantCreation(tier, segmentId, holderHost);
+      holder.getPeon().loadSegment(segment, () -> throttler.unregisterReplicantCreation(tier, segmentId, holderHost));
     }
 
-    return assignedCount;
+    return numToAssign;
   }
 
-  private CoordinatorStats drop(
-      final Map<String, Integer> loadStatus,
+  private void drop(
+      final DruidCoordinatorRuntimeParams params,
       final DataSegment segment,
-      final DruidCoordinatorRuntimeParams params
+      final CoordinatorStats stats
```

Comment:
this is deceiving, it is a final object reference, but the contents are modified. Suggest adding a method comment to such effect.